clients/src/main/java/org/apache/kafka/clients/Metadata.java

```diff
@@ -161,6 +253,40 @@ public synchronized int requestUpdate() {
         return this.version;
     }
 
+    /**
+     * Request an update for the partition metadata iff the given leader epoch is at newer than the last seen leader epoch
+     */
+    public synchronized boolean updateLastSeenEpochIfNewer(TopicPartition topicPartition, int leaderEpoch) {
+        Objects.requireNonNull(topicPartition, "TopicPartition cannot be null");
+        if (updateLastSeenEpoch(topicPartition, leaderEpoch, oldEpoch -> leaderEpoch > oldEpoch) == leaderEpoch) {
+            this.needUpdate = true;
+            return true;
+        } else {
+            return false;
+        }
+    }
+
+    // Visible for testing
+    Optional<Integer> lastSeenLeaderEpoch(TopicPartition topicPartition) {
+        return Optional.ofNullable(lastSeenLeaderEpochs.get(topicPartition));
+    }
+
+    /**
+     * Update the leader epoch for a partition if the given predicate passes. If updated, remove the stale metadata
+     * record from {@link Cluster}.
+     */
+    private int updateLastSeenEpoch(TopicPartition topicPartition, int epoch, Predicate<Integer> epochTest) {
+        Integer oldEpoch = lastSeenLeaderEpochs.get(topicPartition);
+        if (oldEpoch == null || epochTest.test(oldEpoch)) {
+            log.debug("Setting last seen epoch to {} for partition {}", epoch, topicPartition);
+            lastSeenLeaderEpochs.put(topicPartition, epoch);
+            cache.removePartition(topicPartition);
+            return epoch;
+        } else {
```

Comment:
Maybe we could add a trace message in this case which mentions the old and new epochs. That may make debugging easier if there is a bug which introduces an invalid epoch somehow.